Show the "Temporarily Suspended" notice only when deposits are "blocked"

Jinksi opened the pull request:
Fixes #4452

🏁  **To be released in WCPay 4.7.0**

#### Changes proposed in this Pull Request

<!--
Title: A descriptive, yet concise, title.
-->

<!--
Description: Write a brief summary about this PR. As you compose your summary, consider each of these questions and address them if appropriate. Why is this change needed? What does this change do? Were there other solutions you considered? Why did you choose to pursue this solution? Describe any trade-offs you might have had to make.
-->

* This PR implements the new `deposits_blocked` flag provided by WCPay Server to determine whether to show the "Temporarily suspended" notice for both the **Balances** & **Account Details** component of the **Payments → Overview** page.
* Deposits status is now received via the `deposits.status` key via the `/accounts` endpoint.
  (Changed in 2352-gh-Automattic/woocommerce-payments-server)
* Deposits schedule interval is now received via the `deposits.interval` key via the `/accounts` endpoint.
  (Changed in 2352-gh-Automattic/woocommerce-payments-server)
* If the account does not have deposits blocked, and their deposit schedule is set to `manual`, both **Balances** & **Account Details** components will now show the "Manual" schedule. This differs from existing behaviour, where a `manual` deposits schedule would display the "Temporarily suspended" notice.
* A feature flag has been added (`_wcpay_feature_custom_deposit_schedules`) which is **enabled** by default. **If this feature flag is disabled, the existing logic is used** to determine whether or not to show the "Temporarily suspended" notice (it will show if the deposit schedule is set to `manual`).




For the **Balances** component, the `account.deposits_blocked` flag provided by the `deposits/overview-all` endpoint is used.

![image](https://user-images.githubusercontent.com/3147296/182542286-8b40ccc4-fbf0-4d52-976c-836fb12aebbf.png)

For the **Account Details** component, the `deposits.status` property (`blocked`) provided by the `account` endpoint is used.

![image](https://user-images.githubusercontent.com/3147296/182542189-09633a3a-1047-4c91-b444-36b345871be8.png)

<!--
Questions for PR author:
- How can this code break?
- What are we doing to make sure this code doesn't break?
-->

<!--
Images or gifs: Include before and after screenshots or gifs/videos when it makes sense.
-->

#### Testing instructions

<!--
Testing instructions: How should this be tested and how can a reviewer test the end-user functionality? Are there known issues that you plan to address in a future PR? Are there any side effects that readers should be aware of?
-->

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

0. Checkout this wcpay-server branch: `add/2190-get_deposits_status-returns-blocked` (2352-gh-Automattic/woocommerce-payments-server)
   1. Visit the wcpay-server MC page (http://localhost:8086/wp-admin/admin.php?page=wcpay-mc)
   2. View the **Account Tools** by adding a **site ID** to the Blog RC input (I used my WCPay-client blog ID, copied from the [WCPay Dev Tools page](http://localhost:8082/wp-admin/admin.php?page=wcpaydev))
   3. If you had previously Hard/Soft blocked this account, Approve or Clear it using the **Account Tools**.
   4. View the **Stripe Account Dashboard** by clicking the _View in Stripe_ link in the wcpay-server Blog RC. Deposit schedule can be modified by clicking the three-dot menu next to _Account balances_.
5. Test with [WCPay client 4.5.1](https://github.com/Automattic/woocommerce-payments/releases/tag/4.5.1): `git fetch --tags && git checkout 4.5.1`. (Doesn't support `blocked` deposits status)
   1. ⚠️ Make sure you run npm run build:client
   1. Set account deposits schedule to Manual via Stripe dashboard. Clear the account cache contents via WCPay Dev Tools.
   3. **Payments → Overview** should display "Temporarily Suspended" in both the **Balances** and **Account Details** card.
   4. Set account deposits schedule to Daily via Stripe dashboard. Clear the account cache contents via WCPay Dev Tools.
   6. **Payments → Overview** should display "Daily/Automatically" in both the **Balances** and **Account Details** card.
5. Test with WCPay client from this branch. **Feature Flag Enabled**
   1. ⚠️ Make sure you run `npm run build:client`
   2. Set the feature flag option to `'1'`: `update_option( '_wcpay_feature_custom_deposit_schedules', '1');` using [WP-Console](https://wordpress.org/plugins/wp-console/)
   3. Manually change the `woocommerce-payments.php` `Version: 4.7.0` (required for wcpay-server to recognise feature support)
   2. If you had previously Hard/Soft blocked this account, Approve or Clear it using the Account Tools.
   5. Set account deposits schedule to Manual via Stripe dashboard. Clear the account cache contents via WCPay Dev Tools.
   7. **Payments → Overview** should display "Manual" in both the **Balances** and **Account Details** card.
   8. Set account deposits schedule to Daily via Stripe dashboard. Clear the account cache contents via WCPay Dev Tools.
   9. **Payments → Overview** should display "Daily/Automatically" in both the **Balances** and **Account Details** card.
   10. Soft block the account via wcpay-server Blog RC. Clear the account cache contents via WCPay Dev Tools.
   11. **Payments → Overview** should display "Temporarily suspended" in both the **Balances** and **Account Details** card.
4. Test with WCPay client from this branch. **Feature Flag Disabled**
   1. Delete the feature flag option: `delete_option( '_wcpay_feature_custom_deposit_schedules');` using [WP-Console](https://wordpress.org/plugins/wp-console/)
   3. If you had previously Hard/Soft blocked this account, Approve or Clear it using the **Account Tools**.
   4. Set account deposits schedule to Manual via Stripe dashboard. Clear the account cache contents via WCPay Dev Tools.
   7. **Payments → Overview** should display "Temporarily Suspended" in both the **Balances** and **Account Details** card.
   8. Set account deposits schedule to Daily via Stripe dashboard. Clear the account cache contents via WCPay Dev Tools.
   9. **Payments → Overview** should display "Daily/Automatically" in both the **Balances** and **Account Details** card.
   10. Soft block the account via wcpay-server Blog RC. Clear the account cache contents via WCPay Dev Tools.
   11. **Payments → Overview** should display "Temporarily suspended" in both the **Balances** and **Account Details** card.


-------------------

- [x] Run `npm run changelog` to add a changelog file, choose `patch` to leave it empty if the change is not significant. You can add multiple changelog files in one PR by running this command a few times. 
- [x] Covered with tests (or have a good reason not to test in description ☝️)
- [ ] Tested on mobile (or does not apply)
- [x] Check that there is no issue with cached account data when a store upgrades from <4.7.0 (Raised in 2352-gh-Automattic/woocommerce-payments-server#issuecomment-1228018779). Test after 2352-gh-Automattic/woocommerce-payments-server has been deployed.

**Post merge**

<!--
Make sure you edit the page for the current release when adding testing instructions.
We often create a blank page ahead of time for the next release.
If this PR need not be QA tested, edit to 'QA Testing Not Applicable'
-->

- [x] Link to testing instructions from [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) following [these instructions](https://github.com/Automattic/woocommerce-payments/wiki/How-to-write-good-manual-testing-scenarios) :  QA Testing Not Applicable: changes are under-the-hood and require privileged access in order to block/approve accounts etc.
- [x] Add or update [critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Critical-flows) and [testing instructions for critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Testing-instructions-for-critical-flows), if applicable.
- [ ] Add what's changed (description, screenshot, demo videos etc.) to the release announcement post, if applicable.

Jinksi committed:
Add feature flag for custom deposit schedules

Jinksi committed:
DepositsInfo: suspended notice when deposits blocked

If customDepositSchedules flag = true, manual schedule will no longer display suspended notice.

Jinksi committed:
DepositsStatus: suspended notice when deposits blocked

If customDepositSchedules flag enabled, manual schedule will no longer display suspended notice.

Jinksi committed:
Fix error if custom deposits feature flag not set

Jinksi committed:
Test account status comp with custom deposits flag

Jinksi committed:
Update getDepositSchedule unit tests

Add new account.deposits_blocked argument

Test with customDepositSchedules feature flag enabled/disabled

Jinksi commented:
I am leaving this as a draft until there is more certainty of the server-side changes in 2190-gh-Automattic/woocommerce-payments-server.

Jinksi committed:
Add DepositsStatus test for deposits_status = blocked

Jinksi committed:
Add a changelog entry

Jinksi committed:
Merge branch 'develop' into update/4452-suspended-notice-only-when-deposits_status-blocked

Jinksi committed:
Update key used for `/accounts` deposits status

Changed to `deposits.schedule` in wcpay-server PR#2352

Jinksi committed:
Merge branch 'develop' into update/4452-suspended-notice-only-when-deposits_status-blocked

Jinksi committed:
Merge branch 'develop' into update/4452-suspended-notice-only-when-deposits_status-blocked

Jinksi committed:
Merge branch 'develop' into update/4452-suspended-notice-only-when-deposits_status-blocked

Jinksi committed:
Merge branch 'develop' into update/4452-suspended-notice-only-when-deposits_status-blocked

Jinksi committed:
Merge branch 'develop' into update/4452-suspended-notice-only-when-deposits_status-blocked

Jinksi committed:
Accept new accounts endpoint deposits property

Jinksi committed:
DepositsStatus: accept new account deposit prop

The `/accounts` endpoint now includes a `deposits` payload.

`depositsStatus` is now split into `deposits.status` and `deposits.interval`

mattallan reviewed (approved):
Awesome work on this PR @Jinksi and thanks for adding such amazing testing instructions!

I tested and confirmed the following cases:
- :green_circle: Test with WCPay client 4.5.1: `git fetch --tags && git checkout 4.5.1`
  - Confirmed **Payments → Overview** shows "Temporarily Suspended" after setting payout schedule to manual
  - Confirmed **Payments → Overview** shows "Daily/Automatically" after setting payout scheduled to daily
- :green_circle: Test with WCPay client from this branch. Feature Flag Enabled
  - Confirmed **Payments → Overview** now shows "Manual" after setting payout schedule to manual ([screenshot](https://d.pr/Ip1hA0))
  - Confirmed **Payments → Overview** shows "Daily/Automatically" after setting payout scheuled to daily ([screenshot](https://d.pr/02W9h1))
  - Confirmed **Payments → Overview** shows "Temporarily suspended" after soft blocking account. ([screenshot](https://d.pr/3HSjdp))
- :green_circle: Test with WCPay client from this branch. Feature Flag Disabled
  - Confirmed **Payments → Overview** shows "Temporarily Suspended" when payout set to manual and account not blocked
  - Confirmed **Payments → Overview** shows "Daily/Automatically" when payouts set to daily
  - Confirmed **Payments → Overview** shows "Temporarily suspended" when account is soft blocked (checked with payouts schedule set to both daily and manual)

I could see any issues in the code as well so happy to approve this one!

Jinksi committed:
Merge branch 'develop' into update/4452-suspended-notice-only-when-deposits_status-blocked

Jinksi committed:
Rename variable status→interval to clarify meaning

Jinksi commented:
I've resolved merge conflicts related to #4542 in (8eb8087 and 50dba46).

I've tested as per the testing instructions and confirm that all is working as expected ✅

aprea commented on includes/class-wc-payments-features.php:
```diff
@@ -119,6 +119,15 @@ public static function is_documents_section_enabled() {
+	 * @return bool
+	 */
+	public static function is_custom_deposit_schedules_enabled() {
+		return '1' === get_option( '_wcpay_feature_custom_deposit_schedules', '0' );
```
Hey @Jinksi, I think we should default this feature flag to being **on**, i.e. change the second param of `get_option` to `'1'`.

This will make the changes visible as merchants update to 4.7.

The other approach would be to simply remove the feature flag and related code, but this would be a fairly significant change and would require another review and additional testing.

It also means that merchants don't have the ability to "turn off" this functionality in the scenario that we ship a bug.

We can always remove the feature flag in a subsequent release.

What do you think?

Jinksi replied:
Sounds good! I'll update the flag to ensure it is enabled by default 👍

I'm testing the upgrade path 4.6.0 → 4.7.0 to ensure there are no hiccups with cached account data, so I'll test the feature flag is enabled by default also.

Jinksi replied:
Updated in 053fc0b and I'll update the PR title/desc now

Jinksi committed:
Enable custom deposit schedules feat flag by default

This will ensure that custom deposit schedule features will be visible to 4.7.0 users.

Leaving the feature flag in place to allow the feature to be disabled.

Jinksi commented:
I've tested the 4.6.0→4.7.0 upgrade path to ensure there are no hiccups with cached account data and that the feature flag is now enabled by default:
* Installed WCPay **4.6.0** and deleted the feature flag option.
* Soft blocked the account via Blog RC.
* This sets the Stripe deposits schedule to manual and displays _Temporarily suspended_ in 4.6.0.
* WCPay Dev Tools → Accounts cache contents includes `deposits.status` = `blocked`
* Uploaded this PR branch. Manually changed version to **4.7.0** with WP plugin editor.
* Confirmed that Payments → Overview correctly shows Temporarily suspended.
	* This is expected because the `deposits.status` property has been cached as `blocked` in the Account cache contents regardless of WCPay version.
* Cleared the account via Blog RC and set the Stripe deposits schedule to manual.
* Ensured that the Manual status is visible, which confirms the feature flag is enabled by default.
